projects/plugins/jetpack/modules/masterbar/admin-menu/class-wpcom-admin-menu.php

```diff
@@ -216,40 +221,56 @@ public function set_site_card_menu_class( array $menu ) {
 	}
 
 	/**
-	 * Returns the first available upsell nudge.
-	 *
-	 * @return array
+	 * Enqueue the script to async load the upsell nudge.
 	 */
-	public function get_upsell_nudge() {
-		require_lib( 'jetpack-jitm/jitm-engine' );
-		$jitm_engine = new JITM\Engine();
-
-		$message_path = 'calypso:sites:sidebar_notice';
-		$current_user = wp_get_current_user();
-		$user_id      = $current_user->ID;
-		$user_roles   = implode( ',', $current_user->roles );
-		$query_string = array(
-			'message_path' => $message_path,
-		);
-
-		// Get the top message only.
-		$message = $jitm_engine->get_top_messages( $message_path, $user_id, $user_roles, $query_string );
-
-		if ( isset( $message[0] ) ) {
-			$message = $message[0];
-			return array(
-				'content'                      => $message->content['message'],
-				'cta'                          => $message->CTA['message'], // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
-				'link'                         => $message->CTA['link'], // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
-				'tracks_impression_event_name' => $message->tracks['display']['name'],
-				'tracks_impression_cta_name'   => $message->tracks['display']['props']['cta_name'],
-				'tracks_click_event_name'      => $message->tracks['click']['name'],
-				'tracks_click_cta_name'        => $message->tracks['click']['props']['cta_name'],
-				'dismissible'                  => $message->is_dismissible,
-				'feature_class'                => $message->feature_class,
-				'id'                           => $message->id,
+	public function enqueue_async_upsell_nudge() {
+		if ( ! did_action( 'enqueue_block_editor_assets' ) ) {
```

Comment:
Could we use [`wp_script_is`](https://developer.wordpress.org/reference/functions/wp_script_is/) instead? Otherwise this piece of logic will be tied to an external behavior that we don't control (imagine that `enqueue_block_editor_assets` stops enqueing the needed scripts at some point in the future).

Reply:
In general, I found the logic below a bit fragile since it needs to replicate several things that we do currently on `wpcom` for simple sites, and that could change at any point without us noticing that we'd need to also modify this file.

I think we can avoid all of this if the new script is based on the previous approach: make AJAX requests to the site (as opposed to `public-api.wordpress.com`) and use the JITM class in the server side to return the upsell nudges to render.

Something like (untested):

```php
add_action( 'wp_ajax_get_upsell_nudge', array( $this, 'wp_ajax_get_upsell_nudge' ) );

public function wp_ajax_get_upsell_nudge() {
  $nudge = $this->get_upsell_nudge();
  if ( ! $nudge ) {
    return;
  }

  $dismiss_button = '';
  if ( $nudge['dismissible'] ) {
    $dismiss_button = '<svg xmlns="http://www.w3.org/2000/svg" data-feature_class="%1$s" data-feature_id="%2$s" viewBox="0 0 24 24" class="gridicon gridicons-cross dismissible-card__close-icon" height="24" width="24"><g><path d="M18.36 19.78L12 13.41l-6.36 6.37-1.42-1.42L10.59 12 4.22 5.64l1.42-1.42L12 10.59l6.36-6.36 1.41 1.41L13.41 12l6.36 6.36z"></path></g></svg>';
    $dismiss_button = sprintf( $dismiss_button, esc_attr( $nudge['feature_class'] ), esc_attr( $nudge['id'] ) );
  }

 return '
<div class="upsell_banner">
	<div class="banner__info">
		<div class="banner__title">%1$s</div>
	</div>
	<div class="banner__action">
		<button type="button" class="button">%2$s</button>
	</div>%3$s
</div>';
}
```

```js
makeAjaxRequest(
  'GET',
  ajaxurl,
  'action=get_upsell_nudge',
  function( response ) {
    adminMenu.querySelector( '#toplevel_page_site_card' ).insertAdjacentHTML( 'afterend', response );
  }
);
```

That way we won't be tied to the wpcom scripts overrided and won't need to enqueue `jquery.wpcom-proxy-request.js`, `gutenberg-wpcom-apifetch.js`, and friends.

Reply:
Thanks for this. I like the approach and will go ahead with it in another PR.